### DISCUSSION TITLE
快速开始

### DISCUSSION THREAD

**Discussion by lencx** [lencx/gg#5] [Docs]
### Step1：fork 项目

[点击项目](https://github.com/lencx/gg) 右上角的 Fork 按钮，fork 后，可以对仓库进行重命名，修改成自己喜欢的名称。

重命名（可选操作）:

1. 访问 `https://github.com/<owner>/<repo>/settings`
2. `General` -> `Repository name` -> `Rename`

**⚠️ 注意：强烈推荐使用 `fork` 方式，不要对仓库 `main` 分支代码进行改动，这样可以和上游仓库代码始终保持一致，在升级时，可以快速同步，降低个人的维护成本。**

### Step2：开启 Discussions

`Repo -> Settings -> General -> Features -> Discussions`

### Step3：创建配置

在 Discussions 中新增一个标题为 `rgd.yml` 的 issues，并在评论框中输入以下代码：

````md
```yml
### 必须配置
# github 用户名，支持个人或组织
owner: lencx
# github 仓库名称
repo: gg

### 可选配置，如果 type 为 `issues` 类型，才需要以下配置
# type: `issues` | `discussions` | `discussions2`, default is `discussions`
# discussions：表示需要写 discussions 的仓库与 fork 此项目的仓库为同一个
# discussions2：表示需要写 discussions 的仓库与 fork 此项目的仓库彼此独立
type: issues
issues-owner: <github_owner>
issues-repo: <github_repo>
# state: `open` or `closed`, default is `all`
issues-state: closed

### 可选配置，如果 type 为 `discussions2` 类型，才需要以下配置
# type: `issues` | `discussions` | `discussions2`, default is `discussions`
type: discussions2
# 写 discussions 的用户名及仓库名
dis-owner: <github_owner>
dis-repo: <github_repo>

### 域名配置
# 个人域名，如果没有则不需要设置
# 默认使用 github 二级域名，即 https://<owner>.github.io/<repo>
cname: gg.nofwl.com

### website 配置
website:
  # 1. 格式为 png 的 240x240 正方形图片
  # 2. 在线图片 URL 地址，不设置则使用默认 logo
  logo: https://user-images.githubusercontent.com/16164244/162659556-64dd7da8-321f-4525-bbbf-975f02b4800f.png
  # 网站标题
  title: GG
  # 网站描述
  description: A gatsby website builder based on github discussions

### PWA Manifest 配置
# https://www.gatsbyjs.com/plugins/gatsby-plugin-manifest
manifest:
  name: GG
  short_name: GG
  start_url: /
  background_color: '#fafafa'
  theme_color: '#232629'
  display: standalone
  # 1. 格式为 png 的 240x240 正方形图片
  # 2. 在线图片 URL 地址，不设置则使用默认 logo
  icon: https://user-images.githubusercontent.com/16164244/162659556-64dd7da8-321f-4525-bbbf-975f02b4800f.png
```
````

配置参考：

- [issues 类型](https://github.com/fe-cool/hunter/discussions/2)
- [discussions 类型](https://github.com/rwasm/rwasm.github.io/discussions/2)
- [自定义域名](https://github.com/lencx/gg/discussions/2)

### Step4：生成 Token

[生成 Token](https://github.com/settings/tokens)

New personal access token

1. `Note` - 备注随便输入，主要做标识
3. `Expiration：No expiration` - 过期时间看需要，这里选择不过期
4. `Select scopes: workflow` - 只需要勾选 `workflow`
5. 点击 `Generate token` 按钮，然后复制生成好的 `Token`

### Step5：设置 Secrets

`Repo -> Settings -> Secrets -> Actions -> New repository secret`

 
当 `GG_TYPE` 为 `issues` 时，才需要下面的变量配置：

1. GG_TYPE
    - Name: `GG_TYPE`
    - Value: `issues`
2. GG_ISSUES_OWNER：`GG_TYPE` 为 `issues` 时，此字段必须
    - Name: `GG_ISSUES_OWNER`
    - Value: `<github_owner>`
3. GG_ISSUES_USER：`GG_TYPE` 为 `issues` 时，此字段必须
    - Name: `GG_ISSUES_REPO`
    - Value: `<github_repo>`
4. GG_ISSUES_STATE：可选，如果未设置则包含 `OPEN` 和 `CLOSED` 的 issues
    - Name: `GG_ISSUES_STATE`
    - Value: `CLOSED`

当 `GG_TYPE` 为 `discussions2` （部署网站的仓库与 discussions 数据不在一个仓库）时，才需要下面的变量配置：

1. GG_TYPE
    - Name: `GG_TYPE`
    - Value: `discussions2`
2. GG_DIS_OWNER：`GG_TYPE` 为 `discussions2` 时，此字段必须
    - Name: `GG_DIS_OWNER`
    - Value: `<github_owner>`
3. GG_DIS_REPO：`GG_TYPE` 为 `discussions2` 时，此字段必须
    - Name: `GG_DIS_REPO`
    - Value: `<github_repo>`

## 自定义域名

如果没有域名则不需要设置，默认使用 `github` 二级域名，即 `https://<owner>.github.io/<repo>`。

自定义域名以阿里云为例：
<img width="500" alt="CNAME" src="https://user-images.githubusercontent.com/16164244/162999068-e1a7db79-6dfd-4b63-9a9f-3145d7151c37.jpg" />

## 构建

- 提交代码会触发构建（🙅🏻‍♂️不推荐，会和上游仓库产生冲突，无法拉取最新代码）
- 每隔一小时会自动构建一次（💁🏻‍♂️ 默认，首次创建仓库需要先启用 workflow）
- 手动触发构建（🙋🏻‍♂️ 想要部署到网站预览，可以点击 `Re-run all jobs` 按钮来部署）

在 Fork 项目后，点击 `Actions` Tab 项，在 `All workflows` 中选择 `github pages`，*点击 `Enable workflow`*，每隔一小时会自动触发一次构建。当自动构建过一次之后，会在 `github pages` 下产生记录，点击最新一条记录，会看到 `Re-run all jobs` 按钮，点击它就可以手动执行构建。

<img width="680" alt="enable workflow" src="https://user-images.githubusercontent.com/16164244/159691189-ea7c88ea-c0a7-49ba-8128-3da2648a8d1e.png">

<img width="680" alt="workflow" src="https://user-images.githubusercontent.com/16164244/159640201-9a8b6229-861e-4664-8ff8-0b8438c19cf8.png">

<img width="680" alt="workflow" src="https://user-images.githubusercontent.com/16164244/159640200-19d25bad-0b14-4a69-a547-e9584125074c.png">

**Comment by excing** [comment]
你好，Github Action 运行失败，Run yarn build 这里报错了：

```
error There was an error in your GraphQL query:

Cannot query field "childrenDiscussionsJson" on type "File".

If you don't expect "childrenDiscussionsJson" to exist on the type "File" it is most likely a typo.
```

请问，怎么办？

**Reply by lencx** [reply]
你可以在本地 build 项目，也会看到类似错误。方便排查问题，看样子是 graphql 字段出问题了

**Comment by shenweiyan** [comment]
> 如果没有域名则不需要设置，默认使用 github 二级域名，即 `https://<owner>.github.io/<repo>`。

通过 `https://<owner>.github.io/<repo>` 访问后，对应的文章链接都变成了 `https://<owner>.github.io/issues/<issues_number>`，而且头部的 `Category`、`Labels` 这些链接都没法点击了。

示例站点：<https://github.com/shenweiyan/gg-discussions>

**Reply by shenweiyan** [reply]
调整了一下配置，重新部署即恢复正常了！感谢作者优秀的开源项目！
